File: Specializations/Data Structures and Algorithms/Data Structures/Starters PA1/network_packet_processing_simulation/process_packages.py
# python3

class Request:
    def __init__(self, arrival_time, process_time):
        self.arrival_time = arrival_time
        self.process_time = process_time

class Response:
    def __init__(self, dropped, start_time):
        self.dropped = dropped
        self.start_time = start_time

class Buffer:
    def __init__(self, size):
        self.size = size
        self.finish_time_ = []

    def Process(self, request):
        # write your code here
        return Response(False, -1)

def ReadRequests(count):
    requests = []
    for i in range(count):
        arrival_time, process_time = map(int, input().strip().split())
        requests.append(Request(arrival_time, process_time))
    return requests

def ProcessRequests(requests, buffer):
    responses = []
    for request in requests:
        responses.append(buffer.Process(request))
    return responses

def PrintResponses(responses):
    for response in responses:
        print(response.start_time if not response.dropped else -1)

if __name__ == "__main__":
    size, count = map(int, input().strip().split())
    requests = ReadRequests(count)

    buffer = Buffer(size)
    responses = ProcessRequests(requests, buffer)

    PrintResponses(responses)

File: Specializations/Data Structures and Algorithms/Data Structures/Week 1/Assignment/src/main/scala/ProcessPackages.scala
import java.io.{BufferedReader, IOException, InputStream, InputStreamReader}
import java.util.StringTokenizer

import scala.collection.mutable.Queue

object ProcessPackages {
  class Request(val arrivalTime: Int, val processTime: Int)
  class Response(val dropped: Boolean, val startTime: Int)

  class Buffer(val size: Int) {
    val buffer = Queue[Int]()

    def process(request: Request): Response = {
      def freeBuffer(q: Queue[Int]): Unit = {
        if (q.nonEmpty && q.head <= request.arrivalTime) {
          q.dequeue
          freeBuffer(q)
        }
      }

      freeBuffer(buffer)

      // If still full, just drop it
      if (buffer.length == size) new Response(true, -1)
      else {
        val startTime = if (buffer.isEmpty) request.arrivalTime else buffer.last
        buffer.enqueue(startTime + request.processTime)
        new Response(false, startTime)
      }
    }
  }

  object Process {
    def readQueries(requestCount: Int)(read: Int => (Int, Int)): List[Request] = {
      def fillRequests(i: Int, requests: List[Request]): List[Request] = {
        if (i >= requestCount) requests
        else {
          val rawRequest = read(i)
          val request = new Request(rawRequest._1, rawRequest._2)
          fillRequests(i + 1, request :: requests)
        }
      }
      fillRequests(0, Nil)
    }

    def processRequests(requests: List[Request], buffer: Buffer): List[Response] =
      requests map (req => buffer.process(req))

    def printResponses(responses: List[Response]): Unit =
      responses.foreach(res =>
        if (res.dropped) println(-1)
        else println(res.startTime))

    def run(bufferSize: Int, requestCount: Int)(read: Int => (Int, Int)): Unit = {
      val queries = readQueries(requestCount)(read)
      val responses = processRequests(queries.reverse, new Buffer(bufferSize))
      printResponses(responses)
    }
  }

  def main(args: Array[String]): Unit = {
    val scanner: FastScanner = new FastScanner(System.in)
    val size: Int = scanner.nextInt
    val count: Int = scanner.nextInt

    Process.run(size, count)(_ => (scanner.nextInt, scanner.nextInt))
  }

  class FastScanner(val stream: InputStream) {

    var br: BufferedReader = null
    var st: StringTokenizer = null

    try
      br = new BufferedReader(new InputStreamReader(stream))

    catch {
      case e: Exception => {
        e.printStackTrace()
      }
    }

    def next: String = {
      while (st == null || !st.hasMoreTokens)
        try
          st = new StringTokenizer(br.readLine)

        catch {
          case e: IOException => {
            e.printStackTrace()
          }
        }
      st.nextToken
    }

    def nextInt: Int = next.toInt
  }
}
